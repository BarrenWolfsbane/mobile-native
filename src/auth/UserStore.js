import {
  observable,
  action
} from 'mobx';

import channelService from './../channel/ChannelService';
import sessionService from '../common/services/session.service';
import UserModel from './../channel/UserModel';
import { MINDS_FEATURES } from '../config/Config';
import { Alert } from 'react-native';

/**
 * Login Store
 */
class UserStore {
  @observable me = {};

  @action
  setUser(user) {
    if (!user || !user.guid)
      return
    this.me = UserModel.create(user);
  }

  @action
  setRewards(value) {
    this.me.rewards = !!value;
  }

  @action
  setWallet(value) {
    this.me.eth_wallet = value;
  }

  @action
  setTosLastUpdate(value) {
    this.me.last_accepted_tos = value;
  }

  hasRewards() {
    return !!(this.me && this.me.rewards);
  }

  hasEthWallet() {
    return !!(this.me && this.me.eth_wallet);
  }

  hasAvatar() {
    return this.me.icontime !== this.me.time_created;
  }

  @action
  clearUser() {
    this.me = {};
  }

  @action
  async load(refresh = false) {
    if (!refresh) this.me = {};

    let response = await channelService.load('me');

    //if (!response.channel) {
    //  return sessionService.setToken(null);
    //}

    this.setUser(response.channel);
    if (this.me.canCrypto) {
      MINDS_FEATURES.crypto = true;
    }
    return this.me;

  }

  isAdmin() {
    return this.me.admin;
  }

  @action
  reset() {
    this.me = {};
  }

}

export default UserStore;
