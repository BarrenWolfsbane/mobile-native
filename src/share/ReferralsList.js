import React, {
  Component
} from 'react';

import {
  Text,
  FlatList,
  View,
  TouchableOpacity,
  StyleSheet,
} from 'react-native';

import {
  observer,
  inject
} from 'mobx-react/native'

import { CommonStyle } from '../styles/Common';

import i18n from '../common/services/i18n.service';
import CenteredLoading from '../common/components/CenteredLoading';
import ReferralsRow from './ReferralsRow';

@inject('referrals')
@observer
export default class ReferralsList extends Component {

  /**
   * On component will unmount
   */
  componentWillUnmount() {
    //TODO clear list
  }

   /**
   * On component will mount
   */
  componentDidMount() {
    //TODO prepare data and load list
    this.setState(() => {
      this.loadMore();
    });
  }

  /**
   * Render
   */
  render() {
    const referrals = this.props.referrals;
    const entities = referrals.list.entities;
    const header = this.getHeader();
    let view = !referrals.list.loaded ? (<CenteredLoading />) : (<View />);
    return (
      <FlatList
        data={entities.slice()}
        renderItem={this.renderRow}
        keyExtractor={(item, index) => item.timestamp || 0 + index}
        onRefresh={this.refresh}
        refreshing={referrals.list.refreshing}
        onEndReached={this.loadMore}
        // onEndReachedThreshold={0}
        ListHeaderComponent={header}
        ListEmptyComponent={view}
        style={[CommonStyle.flexContainer, CommonStyle.backgroundWhite]}
      />
    )
  }

  /**
   * Get header
   */
  getHeader() {
    return (
      <View style={styles.header}>
        <View style={[CommonStyle.rowJustifyStart, styles.row]}>
          <Text style={[{fontWeight: '800', fontSize: 10, flex:5 }]}>{i18n.t('auth.username')}</Text>
          <Text style={[{fontWeight: '800', fontSize: 10, flex:2 }]}>{i18n.t('referrals.referralsStatus')}</Text>
          <Text style={[{fontWeight: '800', fontSize: 10, flex:3 }]}>{i18n.t('referrals.referralsWalletSingup')}</Text>
        </View>
      </View>
    );
  }

  /**
   * Load more data
   */
  loadMore = () => {
    this.props.referrals.loadList();
  }

  /**
   * Render list's rows
   */
  renderRow = (row, i) => {
    const item = row.item;
    return (
      <ReferralsRow item={item} key={i} navigation={this.props.navigation} />
    )
  }

  /**
   * Refresh list
   */
  refresh = () => {
    this.props.referrals.refresh();
  }
}

const styles = StyleSheet.create({
  header: {
    paddingTop: 5,
  },
  row: {
    paddingTop: 16,
    paddingBottom: 16,
    paddingLeft: 8,
    paddingRight: 8,
    borderBottomWidth: 1,
    borderBottomColor: '#ececec',
  },
  column: {
    color: '#555',
  },
});
