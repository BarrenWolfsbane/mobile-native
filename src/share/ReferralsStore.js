import {
  observable,
  computed,
  action
} from 'mobx';

import OffsetListStore from '../common/stores/OffsetListStore';
import referralsService from './ReferralsService';
import logService from '../common/services/log.service';
import ReferralModel from './ReferralModel';

export default class ReferralsStore {
  list = new OffsetListStore('shallow');
  loading = false;

  /**
   * Load list
   */
  async loadList() {
    try {
      if (this.list.cantLoadMore() || this.loading) {
        return Promise.resolve();
      }
      this.loading = true;
      referrals = await referralsService.getReferrals(this.list.offset);
      referrals.entities = ReferralModel.createMany(referrals.entities);
      this.list.setList(referrals);
      this.loaded = true;
    } catch(err) {
      logService.exception('[ReferralsStore]', err);
    } finally {
      this.loading = false;
    }
  }

  @action
  refresh() {
    this.list.refresh();
    this.loadList()
      .finally(() => {
        this.list.refreshDone();
      });
  }
}
