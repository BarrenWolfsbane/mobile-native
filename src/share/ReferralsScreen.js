import React, {
  Component
} from 'react';

import {
  Text,
  ScrollView,
  View,
  StyleSheet,
  TouchableOpacity,
} from 'react-native';

import { Icon } from 'react-native-elements'

import {
  observer,
  inject
} from 'mobx-react/native'

import { CommonStyle } from '../styles/Common';

import i18n from '../common/services/i18n.service';
import settingsService from '../settings/SettingsService';
import settingsStore from '../settings/SettingsStore';
import shareService from "../share/ShareService";
import ReferralsList from './ReferralsList';

/**
 * Token and Rewards Screen
 */
@inject('user')
@observer
export default class ReferralsScreen extends Component {

  static navigationOptions = ({ navigation }) => {
    return {title: i18n.t('referrals.referralsTitle')}
  }

  /**
   * Render
   */
  render() {
    return (
      <View style={[CommonStyle.flexContainer, CommonStyle.backgroundWhite, CommonStyle.padding]}>
        <Text style={[styles.title, styles.textContent]}>{i18n.t('referrals.referralsHowWork')}</Text>
        <Text style={[CommonStyle.fontS, styles.textContent]}>
          {i18n.t('referrals.referralsDescription')}
        </Text>
        <ReferralsList navigation={this.props.navigation} />
        <Icon
          raised
          name="md-share"
          type='ionicon'
          color='#fff'
          size={32}
          containerStyle={ settingsStore.leftHanded ? styles.leftSide : styles.rightSide }
          onPress={() => this.share()}
        />
      </View>
      
    )
  }

  share = () => {
    setTimeout(() => {
      shareService.invite(this.props.user.me.name);
    }, 100); 
  }

}

const styles = StyleSheet.create({
  title: {
    fontWeight: 'bold',
    fontFamily: 'Roboto',
    marginTop: 8
  },
  textContent: {
    color: '#444',
    padding: 4
  },
  rightSide: {
    position:'absolute',
    backgroundColor:'#1aaa55',
    width:55,
    height:55,
    bottom:8,
    zIndex:1000,
    right:8
  },
  leftSide: {
    position:'absolute',
    backgroundColor:'#1aaa55',
    width:55,
    height:55,
    bottom:8,
    zIndex:1000,
    left:8
  }
});
