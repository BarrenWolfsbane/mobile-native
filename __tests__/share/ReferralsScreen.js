import 'react-native';
import React from 'react';
import { shallow } from 'enzyme';
import ReferralsScreen from '../../src/share/ReferralsScreen';
import UserStore from '../../src/auth/UserStore';
import ReferralsList from '../../src/share/ReferralsList';

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';
import ReferralsStore from '../../src/share/ReferralsStore';

jest.mock('../../src/auth/UserStore');
jest.mock('../../src/share/ReferralsStore');
jest.mock('../../src/share/ReferralsList', () => 'ReferralsList');


/**
 * Tests
 */
describe('referrals screen component', () => {

  let user, store;
  beforeEach(() => {
    user = new UserStore();
    user.me = {
      guid: 'guidguid'
    };
    store = new ReferralsStore();
  });

  it('should renders correctly', () => {

    const component = renderer.create(
      <ReferralsScreen user={user} referrals={store} />
    ).toJSON();

    expect(component).toMatchSnapshot();
  });
});