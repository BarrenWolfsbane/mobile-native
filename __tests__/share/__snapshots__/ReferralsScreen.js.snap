// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`referrals screen component should renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "flex": 1,
      },
      Object {
        "backgroundColor": "white",
      },
      Object {
        "padding": 5,
      },
    ]
  }
>
  <Text
    style={
      Array [
        Object {
          "fontFamily": "Roboto",
          "fontWeight": "bold",
          "marginTop": 8,
        },
        Object {
          "color": "#444",
          "padding": 4,
        },
      ]
    }
  >
    How does it work?
  </Text>
  <Text
    style={
      Array [
        Object {
          "fontSize": 12,
        },
        Object {
          "color": "#444",
          "padding": 4,
        },
      ]
    }
  >
    If your friends signs up for Minds within 24 hours of clicking the link you shared with them, they’ll be added to your pending referrals. Once they sign up for the rewards program by setting up their Minds wallet, the referral is complete and you’ll both get +50 added to your contribution scores!
  </Text>
  <ReferralsList />
  <View
    accessible={true}
    isTVSelectable={true}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        Object {
          "margin": 7,
        },
        Object {
          "borderRadius": 36,
          "height": 68,
          "width": 68,
        },
        Object {
          "shadowColor": "rgba(0,0,0, .4)",
          "shadowOffset": Object {
            "height": 1,
            "width": 1,
          },
          "shadowOpacity": 1,
          "shadowRadius": 1,
        },
        Object {
          "alignItems": "center",
          "backgroundColor": "white",
          "justifyContent": "center",
        },
        Object {
          "backgroundColor": "#1aaa55",
          "bottom": 8,
          "height": 55,
          "position": "absolute",
          "right": 8,
          "width": 55,
          "zIndex": 1000,
        },
      ]
    }
  >
    <Text
      allowFontScaling={false}
      style={
        Array [
          Object {
            "color": "#fff",
            "fontSize": 32,
          },
          Array [
            Object {
              "backgroundColor": "transparent",
            },
            undefined,
          ],
          Object {
            "fontFamily": "Ionicons",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          Object {},
        ]
      }
    >
      
    </Text>
  </View>
</View>
`;
